python/ray/air/tests/test_dataset_config.py

```diff
@@ -398,6 +399,111 @@ def check_error(shard, results):
+            ),
+            dataset_config=DatasetConfig(
+                randomize_block_order=False,
+                max_object_store_memory_fraction=max_object_store_memory_fraction,
+                per_epoch_preprocessor=BatchMapper(multiply, batch_format="pandas"),
+            ),
+        )
+
+        def checker(shard, results):
+            assert len(results[0]) == 5, (max_object_store_memory_fraction, results)
+            if max_object_store_memory_fraction is None:
+                assert results[0] == results[1], (
+                    max_object_store_memory_fraction,
+                    results,
+                )
+            else:
+                # Windowed pipelined ingest also reapplies the base
+                # preprocessor on every epoch, so we get a random dataset each
+                # time.
+                assert results[0] != results[1], (
+                    max_object_store_memory_fraction,
+                    results,
+                )
+            assert all(x % 2 == 0 for x in results[0]), (
+                max_object_store_memory_fraction,
+                results,
+            )
+
+        TestStream.train_loop_per_worker(it, checker)
+
+
+def test_nondeterministic_per_epoch_preprocessor(ray_start_4_cpus):
+    ds = ray.data.range_table(5)
+
+    # Use randomized per-epoch preprocessor to check that it gets applied once
+    # per epoch.
+    def rand(x):
+        return x * random.random()
+
+    for max_object_store_memory_fraction in [None, 1, 0.3]:
+        it = make_local_dataset_iterator(
+            ds,
+            preprocessor=None,
+            dataset_config=DatasetConfig(
+                randomize_block_order=False,
+                max_object_store_memory_fraction=max_object_store_memory_fraction,
+                per_epoch_preprocessor=BatchMapper(rand, batch_format="pandas"),
+            ),
+        )
+
+        def checker(shard, results):
+            assert len(results[0]) == 5, (max_object_store_memory_fraction, results)
+            # Per-epoch preprocessor is randomized, so we should get a random
+            # dataset on each epoch.
+            assert results[0] != results[1], (max_object_store_memory_fraction, results)
+
+        TestStream.train_loop_per_worker(it, checker)
+
+
+def test_validate_per_epoch_preprocessor(ray_start_4_cpus):
+    ds = ray.data.range_table(5)
+
+    def multiply(x):
+        return x * 2
+
+    dataset_config = DatasetConfig(
+        per_epoch_preprocessor=BatchMapper(multiply, batch_format="pandas")
+    )
+    DatasetConfig.validated(
+        {
+            "train": dataset_config,
+        },
+        {"train": ds},
+    )
+
+    with pytest.raises(ValueError):
+        dataset_config = DatasetConfig(per_epoch_preprocessor=multiply)
+        DatasetConfig.validated(
+            {
+                "train": dataset_config,
+            },
+            {"train": ds},
+        )
+
```

Comment:
Nit: could you add a comment here and maybe on line 487 describing why `DatasetConfig` raises a `ValueError`? It wasn't obvious to me why `Preprocessor()` is invalid from the reading the test.